File: src/main/java/com/rotpaddon/exampleaddon/power/impl/nonstand/type/tutorial/TutorialPowerType.java
package com.rotpaddon.exampleaddon.power.impl.nonstand.type.tutorial;

import com.github.standobyte.jojo.power.impl.nonstand.INonStandPower;
import com.github.standobyte.jojo.power.impl.nonstand.type.NonStandPowerType;
import com.github.standobyte.jojo.power.impl.stand.IStandPower;
import com.rotpaddon.exampleaddon.action.non_stand.TutorialAction;

import net.minecraft.entity.LivingEntity;
import net.minecraft.potion.EffectInstance;
import net.minecraft.potion.Effects;

public class TutorialPowerType extends NonStandPowerType<TutorialData> {
    public TutorialPowerType(TutorialAction[] startingAttacks, TutorialAction[] startingAbilities) {
        super(startingAttacks, startingAbilities, startingAttacks[0], TutorialData::new);
    }

    public TutorialPowerType(TutorialAction[] startingAttacks, TutorialAction[] startingAbilities, TutorialAction defaultMMB) {
        super(startingAttacks, startingAbilities, defaultMMB, TutorialData::new);
    }

    @Override
    public float tickEnergy(INonStandPower power) {
        if (power.getEnergy() < power.getMaxEnergy()) {
            return power.getEnergy() + 1f;
        }
        return power.getEnergy();
    }

    @Override
    public float getMaxEnergy(INonStandPower power) {
        return 1000;
    }

    @Override
    public boolean isReplaceableWith(NonStandPowerType<?> newType) {
        return true;
    }

    @Override
    public boolean keepOnDeath(INonStandPower power) {
        return true;
    }

    @Override
    public void tickUser(LivingEntity entity, INonStandPower power) {
    	if (!entity.level.isClientSide()) {
    		power.getTypeSpecificData(this).ifPresent(data -> {
    			if (data.isArmored) {
    				entity.addEffect(new EffectInstance(Effects.DAMAGE_RESISTANCE, 25, 255, false, false, true));
    				entity.addEffect(new EffectInstance(Effects.FIRE_RESISTANCE, 25, 255, false, false, true));
    				power.setEnergy(power.getEnergy() - 1f);
    			}
    		});
    	}
    }

    @Override
    public float getTargetResolveMultiplier(INonStandPower power, IStandPower attackingStand) {
        return 0;
    }

    @Override
    public boolean isLeapUnlocked(INonStandPower power) {
        return true;
    }

    @Override
    public float getLeapStrength(INonStandPower power) {
        return 4;
    }

    @Override
    public int getLeapCooldownPeriod() {
        return 20;
    }

    @Override
    public float getLeapEnergyCost() {
        return 100;
    }
}

File: src/main/kotlin/com/rotpaddon/exampleaddon/power/impl/nonstand/type/tutorial/TutorialPowerType.kt
package com.rotpaddon.exampleaddon.power.impl.nonstand.type.tutorial

import com.github.standobyte.jojo.power.impl.nonstand.INonStandPower
import com.github.standobyte.jojo.power.impl.nonstand.type.NonStandPowerType
import com.github.standobyte.jojo.power.impl.stand.IStandPower
import com.rotpaddon.exampleaddon.action.non_stand.TutorialAction
import net.minecraft.entity.LivingEntity
import net.minecraft.potion.EffectInstance
import net.minecraft.potion.Effects

class TutorialPowerType(
    startingAttacks: Array<TutorialAction>,
    startingAbilities: Array<TutorialAction>,
    defaultMMB: TutorialAction
) : NonStandPowerType<TutorialData>(
    startingAttacks,
    startingAbilities,
    defaultMMB,
    ::TutorialData
) {
    constructor(
        startingAttacks: Array<TutorialAction>,
        startingAbilities: Array<TutorialAction>
    ) : this(startingAttacks,startingAbilities,startingAttacks[0])

    override fun tickEnergy(power: INonStandPower): Float {
        return if (power.energy < power.maxEnergy) {
            power.energy + 1f
        } else {
            power.energy
        }
    }

    override fun getMaxEnergy(power: INonStandPower): Float {
        return 1000f
    }

    override fun isReplaceableWith(newType: NonStandPowerType<*>): Boolean {
        return true
    }

    override fun keepOnDeath(power: INonStandPower): Boolean {
        return true
    }

    override fun tickUser(entity: LivingEntity, power: INonStandPower) {
        if (!entity.level.isClientSide) {
            power.getTypeSpecificData(this).ifPresent { data ->
                if (data.isArmored) {
                    entity.addEffect(EffectInstance(Effects.DAMAGE_RESISTANCE, 25, 255, false, false, true))
                    entity.addEffect(EffectInstance(Effects.FIRE_RESISTANCE, 25, 255, false, false, true))
                    power.energy = power.energy - 1f
                }
            }
        }
    }

    override fun getTargetResolveMultiplier(power: INonStandPower, attackingStand: IStandPower): Float {
        return 0f
    }

    override fun isLeapUnlocked(power: INonStandPower): Boolean {
        return true
    }

    override fun getLeapStrength(power: INonStandPower): Float {
        return 4f
    }

    override fun getLeapCooldownPeriod(): Int {
        return 20
    }

    override fun getLeapEnergyCost(): Float {
        return 100f
    }
}